src/Tool/BinaryDecoder/BinaryDecoder.cpp

```diff
@@ -0,0 +1,96 @@
+#include <fstream>
+#include <iostream>
+#include <string>
+
+#include <llvm/Support/CommandLine.h>
+
+llvm::cl::opt<std::string> Filename(
+    llvm::cl::Positional, llvm::cl::desc("<input file>"), llvm::cl::Required);
+llvm::cl::opt<int64_t> Start("s",
+    llvm::cl::desc("Specify the index of the starting byte"),
+    llvm::cl::value_desc("start"), llvm::cl::Required);
+llvm::cl::opt<int64_t> Size("n",
+    llvm::cl::desc("Specify the number of bytes of data to decode"),
+    llvm::cl::value_desc("size"), llvm::cl::Required);
+llvm::cl::opt<bool> Remove(
+    "rm", llvm::cl::desc(
+              "Whether to remove the file being decoded after inspection."));
+
+enum OnnxDataType {
+  UNDEFINED = 0,
+  // Basic types.
+  FLOAT = 1,  // float
+  UINT8 = 2,  // uint8_t
+  INT8 = 3,   // int8_t
+  UINT16 = 4, // uint16_t
+  INT16 = 5,  // int16_t
+  INT32 = 6,  // int32_t
+  INT64 = 7,  // int64_t
+  STRING = 8, // string
+  BOOL = 9,   // bool
+
+  // IEEE754 half-precision floating-point format (16 bits wide).
+  // This format has 1 sign bit, 5 exponent bits, and 10 mantissa bits.
+  FLOAT16 = 10,
+
+  DOUBLE = 11,
+  UINT32 = 12,
+  UINT64 = 13,
+  COMPLEX64 = 14,  // complex with float32 real and imaginary components
+  COMPLEX128 = 15, // complex with float64 real and imaginary components
+
+  // Non-IEEE floating-point format based on IEEE754 single-precision
+  // floating-point number truncated to 16 bits.
+  // This format has 1 sign bit, 8 exponent bits, and 7 mantissa bits.
+  BFLOAT16 = 16,
+
+  // Future extensions go here.
+};
```

Comment:
Would we be able to gen this from the gen_onnx_mlir scripts? Just thinking that I would not want to have to update such info in lots of distinct places?